fission-web-server/library/Fission/Web/Server/Types.hs

```diff
@@ -140,6 +155,106 @@ instance MonadDB (Transaction Server) Server where
+              , rrset_comments = Nothing
+              }
+      rrsets = RRSets [rrset]
+    resp <- liftIO $ runClientM (PDNS.updateRecords "localhost" zoneTxt rrsets) clientEnv
+    case resp of
+      Left err -> do
+        logWarn @Text "PowerDNS.set failed"
+        return . Left $ Web.Error.toServerError err
+      Right good -> do
+        logDebug @Text "PowerDNS.set succeeded"
+        return $ Right good
+    where
+      formatRecords :: NonEmpty Text -> [Record]
+      formatRecords values =
+        fmap newRecord (toList values)
+
+      newRecord :: Text -> Record
+      newRecord content =
+        Record { record_content = format recType content, record_disabled = False }
+
+      format :: PowerDNS.API.Zones.RecordType -> Text -> Text
+      format TXT = UTF8.wrapIn "\""
+      format _   = identity
+
+  get url _ = do
+    logDebug $ "PowerDNS.get for " <> textDisplay url
+    clientEnv <- pdnsClientEnv
+    resp <- liftIO $ runClientM (PDNS.search "localhost" (textDisplay url) 1 (Just TyRecord)) clientEnv
+    case resp of
+      Left err -> do
+        logWarn @Text "PowerDNS.get failed"
+        return . Left $ Web.Error.toServerError err
+      Right good -> do
+        logDebug @Text "PowerDNS.get succeeded"
+        return $ Right good
+
+  clear url zoneID = do
+    logDebug $ "PowerDNS.clear " <> textDisplay url
+    results <- PowerDNS.get url zoneID
+    case results of
+      Left err -> do
+        return $ Left err
+      Right searchResults -> do
+        mapM_ removeResult searchResults
+        return $ Right ()
+    where
+      removeResult :: (MonadIO m, MonadReader Config m, MonadThrow m) => PDNS.SearchResult -> m ()
+      removeResult searchResult = do
+        clientEnv <- pdnsClientEnv
+        let
+          recType = convertType $ PDNS.sr_type searchResult
+          rrset =
+            RRSet { rrset_name = PDNS.sr_name searchResult
+                  , rrset_type = recType
+                  , rrset_ttl = 10
+                  , rrset_changetype = Just PowerDNS.API.Zones.Delete
+                  , rrset_records = Nothing
+                  , rrset_comments = Nothing
+                  }
+          rrsets = RRSets [rrset]
+        _ <- liftIO $ runClientM (PDNS.updateRecords "localhost" zoneID rrsets) clientEnv
+        return ()
+
+      convertType :: Text -> PowerDNS.API.Zones.RecordType
+      convertType "CNAME" = CNAME
+      convertType "TXT"   = TXT
+      convertType _       = PowerDNS.API.Zones.A
+
+pdnsClientEnv :: (MonadReader Config m, MonadThrow m, MonadIO m) => m ClientEnv
+pdnsClientEnv = do
+  apiUrl <- asks pdnsURL
+  apiKey <- asks pdnsApiKey
+  uri <- parseBaseUrl $ Text.unpack (textDisplay apiUrl)
+  let
+    rawHTTPSettings =
+      case baseUrlScheme uri of
+        Http  -> defaultManagerSettings
+        Https -> tlsManagerSettings
+
+  mgr <- liftIO $ newManager rawHTTPSettings
+  return $ PDNS.applyXApiKey (textDisplay apiKey) (mkClientEnv mgr uri)
```

Comment:
This function should get called once in `Main.hs`, and then the resulting `rawHttpSettings` get threaded through the code via `Server`'s internal `Config`. You _may_ be able to reuse on of these:

https://github.com/fission-suite/fission/blob/c7f5907867e94b76bdcbed3a499a4128263d7bfd/fission-web-server/library/Fission/Web/Server/Config/Types.hs#L35-L36

Which would look something like:

```haskell
do
  -- ...
  httpMgr  <- asks httpManager
  httpsMgr <- asks tlsManager
```

Reply:
(`asks` pulls values out of the `Config` contained in `Server`. This is the `MonadReader Config Server` class instance AKA "it knows how to access the ambient configured values without explicitly threading them around everywhere)

Reply:
🤦  I added a `pdnsManager` bit to config and then when I refactored this function I forgot to actually *use* it. Fixed now, I think?